graph/path/shortest.go

```diff
@@ -103,6 +109,14 @@ func (p *Shortest) add(u graph.Node) int {
 func (p Shortest) set(to int, weight float64, mid int) {
 	p.dist[to] = weight
 	p.next[to] = mid
+	if p.cycCosts != nil {
```

Comment:
I think this can be optimised by requiring `p.cycCosts != nil && weight < 0`, since only negative weights will ever lead to a negative cycle, so we should only bother storing things when they have reached negativity.

Reply:
Actually, it can just be `weight < 0`, since if we have the possibility of negative weights (we are using Bellman-Ford), we have already made the map.